File: objC/ddutils-common/model/DDRequestQueue [ios+osx]/DDRequestQueue.m
//
//  DDRequestQueue.m
//  myAudi
//
//  Created by Dominik Pich on 2/1/16.
//  Copyright © 2016 Sapient GmbH. All rights reserved.
//
#import <SystemConfiguration/SystemConfiguration.h>

#import "DDRequestQueue.h"
#if DDREQUESTQUE_USES_URL_SESSION == 2
#import <AFNetworking/AFNetworking.h>
#endif

@interface DDRequestQueue ()

@property(readwrite) NSString *name;
@property(readwrite) BOOL persistent;
@property(readwrite) NSArray *queuedRequests;
@property(readwrite) BOOL queueRunning;
@property(assign) SCNetworkReachabilityRef currentReachability;
@end

@implementation DDRequestQueue

#define kDefaultName @"defaultQueue"
#define kDefaultIsPersistent YES
#define kDataKey @"Data"

+ (instancetype)defaultQueue {
    static dispatch_once_t onceToken;
    static id _defaultQueue;
    dispatch_once(&onceToken, ^{
        _defaultQueue = [[[self class] alloc] initWithName:kDefaultName persistent:kDefaultIsPersistent];
    });
    return _defaultQueue;
}

- (id)initWithName:(NSString*)name persistent:(BOOL)persistent {
    self = [super init];
    if(self) {
        self.suspended = YES;
        self.name = name;
        self.persistent = persistent;
        self.queuedRequests = [NSMutableArray array];
        
        if(self.persistent) {
            BOOL br = [self load];
            NSAssert(br, @"Failed to load persistent queue");
            
            if(!br) {
                @throw [NSException exceptionWithName:@"Internal" reason:@"Failed to load persistent queue" userInfo:nil];
            }
        }
    }
    return self;
}

- (BOOL)enqueueRequest:(NSURLRequest*)request {
    NSParameterAssert(request);
    
    BOOL oldSuspended = self.suspended;
    self.suspended = YES;

    [(NSMutableArray*)self.queuedRequests addObject:request];
    
    BOOL br = [self.queuedRequests containsObject:request];
    NSAssert(br, @"Failed to enqueue the object");
    
    if(br) {
        if(self.persistent) {
            br = [self save];
            NSAssert(br, @"Failed to persist queue after enqueing request: %@", request);
        }
    }
    
    if(br) {
        self.suspended = oldSuspended;
    }
    
    return br;
}

- (BOOL)cancelRequest:(NSURLRequest*)request {
    NSParameterAssert(request);
    
    BOOL oldSuspended = self.suspended;
    self.suspended = YES;
    
    BOOL br = [self.queuedRequests containsObject:request];
    NSAssert(br, @"request should be enqueued");
    
    if(br) {
        [(NSMutableArray*)self.queuedRequests removeObject:request];
        br = ![self.queuedRequests containsObject:request];
        NSAssert(br, @"request shouldnt be enqueued anymore");
    }
    
    if(br) {
        if(self.persistent) {
            br = [self save];
            NSAssert(br, @"Failed to persist queue after canceling request: %@", request);
        }
    }
    
    self.suspended = oldSuspended;
    
    return br;
}

- (BOOL)cancelAllRequests {
    BOOL oldSuspended = self.suspended;
    self.suspended = YES;

    [(NSMutableArray*)self.queuedRequests removeAllObjects];
    BOOL br = self.queuedRequests.count == 0;
    NSAssert(br, @"no requests shouldnt be enqueued anymore");

    if(br) {
        if(self.persistent) {
            br = [self save];
            NSAssert(br, @"Failed to persist queue after canceling all requests");
        }
    }
    
    self.suspended = oldSuspended;
    
    return br;
}

- (void)setSuspended:(BOOL)suspended {
    _suspended = suspended;
    if(!_suspended && !self.queueRunning)
        [self doNextRequest];
}

#pragma mark - 

-(void)doNextRequest {
    //break if suspended
    if(self.suspended) {
        return;
    }
    
    __weak typeof(self) wself = self;

    //go to main thread
    if(![NSThread isMainThread]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [wself doNextRequest];
        });
    }
    
    //dequeue first
    NSMutableArray *requests = (NSMutableArray*)self.queuedRequests;
    NSURLRequest *request = [requests firstObject];
    if(!request) {
        return;
    }
    [requests removeObjectAtIndex:0];
    
    self.queueRunning = YES;
#if DDREQUESTQUE_USES_URL_SESSION == 1
    NSURLSession *session = [NSURLSession sharedSession];
    [[session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        [wself didFinishRequest:request
                   withResponse:response
                  includingData:data
                       andError:error];
    }] resume];
#elif DDREQUESTQUE_USES_URL_SESSION == 2
//    NSLog(@"+ start: %@", request.URL);
    AFHTTPRequestOperation *operation = [[AFHTTPRequestOperation alloc]
                                         initWithRequest:request];
    [operation setCompletionBlockWithSuccess:^(AFHTTPRequestOperation *operation, id responseObject) {
        [wself didFinishRequest:request
                   withResponse:operation.response
                  includingData:operation.responseData
                       andError:nil];
    } failure:^(AFHTTPRequestOperation * _Nonnull operation, NSError * _Nonnull error) {
        [wself didFinishRequest:request
                   withResponse:operation.response
                  includingData:operation.responseData
                       andError:error];
    }];
    [operation start];
    
#else
    [NSURLConnection sendAsynchronousRequest:request
                                       queue:[NSOperationQueue mainQueue]
                           completionHandler:^(NSURLResponse * _Nullable response, NSData * _Nullable data, NSError * _Nullable connectionError) {
                               [wself didFinishRequest:request
                                          withResponse:response
                                         includingData:data
                                              andError:connectionError];
                           }];
#endif
}

- (void)didFinishRequest:(NSURLRequest*)request
            withResponse:(NSURLResponse*)response
           includingData:(NSData*)data
                andError:(NSError*)error {
    //go to main thread if needed
    if(![NSThread isMainThread]) {
        __weak typeof(self) wself = self;
        dispatch_async(dispatch_get_main_queue(), ^{
            [wself didFinishRequest:request withResponse:response includingData:data andError:error];
        });
        return;
    }
    
    //handle network error
    if(error && [self isNetworkRelatedError:error]) {
//        NSLog(@"+ Will Retry when network is up: %@", request.URL);
        
        //prepare to retry the request
        [(NSMutableArray*)self.queuedRequests insertObject:request atIndex:0];
        
        //wait for network coming back
        [self beginObservingReachabilityStatusForHost:request.URL.host];
        
        return;
    }

//    NSLog(@"+ Did finish: %@", request.URL);
    
    self.queueRunning = NO;

    //we are done with this one, save the array
    if(self.persistent) {
        [self save];
        //errors ignored
    }
    
    //tell our subclass and the delegate everything else
    [self dequeuedRequest:request withResponse:response includingData:data andError:error];
    
    //go on
    [self doNextRequest];
}

- (BOOL)isNetworkRelatedError:(NSError*)error {
    if ([[error domain] isEqualToString:NSURLErrorDomain]){
        switch ([error code]){
            case NSURLErrorInternationalRoamingOff:
            case NSURLErrorCallIsActive:
            case NSURLErrorDataNotAllowed:
            case NSURLErrorNotConnectedToInternet:
                return YES;
        }
    }
    return NO;
}

- (void)dequeuedRequest:(NSURLRequest*)request withResponse:(NSURLResponse*)response includingData:(NSData*)data andError:(NSError*)error {
    id<DDRequestQueueDelegate> delegate = self.delegate;
    if([delegate respondsToSelector:@selector(requestQueue:dequeuedRequest:withResponse:includingData:andError:)]) {
        [delegate requestQueue:self dequeuedRequest:request withResponse:response includingData:data andError:error];
    }
}

#pragma mark -

- (BOOL)load {
    NSURL *docURL = [self URLForPersistentQueue];
    BOOL br = YES;
    
    //if it doesnt exist, quit
    //if it does, read it
    if([[NSFileManager defaultManager] fileExistsAtPath:docURL.path]) {
        NSData *codedData = [[NSData alloc] initWithContentsOfURL:docURL];
        NSKeyedUnarchiver *unarchiver = [[NSKeyedUnarchiver alloc] initForReadingWithData:codedData];
        NSArray *array = [unarchiver decodeObjectForKey:kDataKey];
        [unarchiver finishDecoding];
        
        NSMutableArray *requests = (NSMutableArray*)self.queuedRequests;
        [requests removeAllObjects];
        if(array.count) {
            [requests addObjectsFromArray:array];
        }
    }

    return br;
}

- (BOOL)save {
    NSArray *array = self.queuedRequests;
    NSURL *docURL = [self URLForPersistentQueue];
    BOOL br = YES;

    //if the array is empty, delete the doc if it exists
    //if the array isnt empty, write it

    if(array.count == 0) {
        if([[NSFileManager defaultManager] fileExistsAtPath:docURL.path]) {
            br = [[NSFileManager defaultManager] removeItemAtURL:docURL error:nil];
        }
    }
    else {
        NSMutableData *data = [[NSMutableData alloc] init];
        NSKeyedArchiver *archiver = [[NSKeyedArchiver alloc] initForWritingWithMutableData:data];
        [archiver encodeObject:array forKey:kDataKey];
        [archiver finishEncoding];
        
        br = [data writeToURL:docURL atomically:YES];
    }
    return br;
}

- (NSURL*)URLForPersistentQueue {
    NSURL *library = [[NSFileManager defaultManager] URLForDirectory:NSCachesDirectory inDomain:NSUserDomainMask appropriateForURL:nil create:YES error:nil];
    NSURL *appDir = [library URLByAppendingPathComponent:@"DDRequestQueue"];
    [[NSFileManager defaultManager] createDirectoryAtURL:appDir withIntermediateDirectories:YES attributes:nil error:nil];
    NSString *filename = [@"QueuedRequests-" stringByAppendingString:self.name];
    
    return [appDir URLByAppendingPathComponent:filename];
}

#pragma mark - 

- (void) beginObservingReachabilityStatusForHost:(NSString *)host {
    NSAssert(self.currentReachability==NULL, @"The current reachability should be NULL");
    
    SCNetworkReachabilityRef reachabilityRef = NULL;
    
    void (^callbackBlock)(SCNetworkReachabilityFlags) = ^(SCNetworkReachabilityFlags flags) {
        BOOL reachable = (flags & kSCNetworkReachabilityFlagsReachable) != 0;
        [self host:host didBecomeReachable:reachable];
    };
    
    SCNetworkReachabilityContext context = {
        .version = 0,
        .info = (void *)CFBridgingRetain(callbackBlock),
        .release = CFRelease
    };
    
    
    if ([host length] > 0){
        reachabilityRef = SCNetworkReachabilityCreateWithName(kCFAllocatorDefault, [host UTF8String]);
        if (SCNetworkReachabilitySetCallback(reachabilityRef, ReachabilityCallback, &context)){
            if (!SCNetworkReachabilitySetDispatchQueue(reachabilityRef, dispatch_get_main_queue()) ){
                // Remove our callback if we can't use the queue
                SCNetworkReachabilitySetCallback(reachabilityRef, NULL, NULL);
            }
            self.currentReachability = reachabilityRef;
        }
    }
}


static void ReachabilityCallback(SCNetworkReachabilityRef __unused target, SCNetworkConnectionFlags flags, void* info) {
    void (^callbackBlock)(SCNetworkReachabilityFlags) = (__bridge id)info;
    callbackBlock(flags);
}


- (void) host:(NSString *)__unused host didBecomeReachable:(BOOL)reachable {
    if (reachable){
        [self doNextRequest];
        
        SCNetworkReachabilitySetDispatchQueue(self.currentReachability, NULL);
        CFRelease(self.currentReachability);
        self.currentReachability = NULL;
    }
}

@end

File: swift/ddutils-common/model/DDRequestQueue [ios+osx]/DDRequestQueue.swift
//
//  DDRequestQueue.m
//  myAudi
//
//  Created by Dominik Pich on 2/1/16.
//  Copyright © 2016 Sapient GmbH. All rights reserved.
//
import Foundation
import SystemConfiguration

@objc protocol DDRequestQueueDelegate {
    //called after done with a request for good or for bad -- this EXCLUDES network errors AND cancellations
    func requestQueueDequeuedRequest(queue:DDRequestQueue, request:NSURLRequest, response:NSURLResponse?, data:NSData?, error:NSError?) -> Void

    
    optional func requestQueueWillStartRequest(queue:DDRequestQueue, request:NSURLRequest) -> Void
    optional func requestQueueBeginWaitingForNetwork(queue:DDRequestQueue, host:String) -> Void
    optional func requestQueueEndWaitingForNetwork(queue:DDRequestQueue, host:String) -> Void
}

///when created/loaded from disk, the queue is suspended by default
///NOTE: should be used on main thread
class DDRequestQueue : NSObject {
    
    static let kDefaultName = "defaultQueue"
    static let kDefaultIsPersistent = true
    static let kDataKey = "Data"
    static let defaultQueue = DDRequestQueue(name: kDefaultName, persistent: kDefaultIsPersistent)
    
    var suspended = true {
        didSet {
            if(!suspended && !self.queueRunning) {
                self.doNextRequest();
            }
        }
    }
    
    private(set) var name = ""
    private(set) var persistent = false
    private(set) var queuedRequests = [NSURLRequest]()
    var delegate: DDRequestQueueDelegate?

    private var queueRunning = false
    private var currentReachability: SCNetworkReachabilityRef!;
    
    init(name:String, persistent:Bool) {
        super.init()

        self.name = name
        self.persistent = persistent
        
        if(self.persistent) {
            self.loadList()
        }
    }
    
    func enqueueRequest(request:NSURLRequest) -> Bool {
        let oldSuspended = self.suspended;
        self.suspended = true;
        
        self.queuedRequests.append(request)
        
        var br = (self.queuedRequests.indexOf(request) != nil);
        //        NSAssert(br, @"Failed to enqueue the object");
        
        if(br) {
            if(self.persistent) {
                br = self.saveList();
                //                NSAssert(br, @"Failed to persist queue after enqueing request: %@", request);
            }
        }
        
        if(br) {
            self.suspended = oldSuspended;
        }
        
        return br;
    }
    
    func cancelRequest(request:NSURLRequest) -> Bool {
        let oldSuspended = self.suspended;
        self.suspended = true;
        
        var idx = self.queuedRequests.indexOf(request)
        var br = (idx != nil);
        //        NSAssert(br, @"request should be enqueued");
        
        if let prevIdx = idx where br {
            self.queuedRequests.removeAtIndex(prevIdx)
            idx = self.queuedRequests.indexOf(request)
            br = (idx == nil)
            //            NSAssert(br, @"request shouldnt be enqueued anymore");
        }
        
        if(br) {
            if(self.persistent) {
                br = self.saveList();
                //                NSAssert(br, @"Failed to persist queue after canceling request: %@", request);
            }
        }
        
        self.suspended = oldSuspended;
        
        return br;
    }
    
    func cancelAllRequests() -> Bool {
        let oldSuspended = self.suspended;
        self.suspended = true;
        
        var br = self.queuedRequests.count != 0;
        
        if(br) {
            self.queuedRequests.removeAll()
            br = self.queuedRequests.count == 0;
        }
        //    NSAssert(br, @"no requests shouldnt be enqueued anymore");
        
        if(br) {
            if(self.persistent) {
                br = self.saveList();
                //                NSAssert(br, @"Failed to persist queue after canceling all requests");
            }
        }
        
        self.suspended = oldSuspended;
        
        return br;
    }
    
    // MARK: processing requests
    
    private func doNextRequest() -> Void {
        //break if suspended
        if(self.suspended) {
            return;
        }
        
        //go to main thread
        if(!NSThread.isMainThread()) {
            dispatch_async(dispatch_get_main_queue()) {
                self.doNextRequest()
            }
        }
        
        //dequeue first
        if(self.queuedRequests.count == 0) {
            return;
        }
        let request = self.queuedRequests[0]
        self.queuedRequests.removeAtIndex(0)
        
        self.queueRunning = true;
        
        self.requestQueueWillStartRequest(request)
        
        //hand it off to ios
        let session = NSURLSession.sharedSession()
        session.dataTaskWithRequest(request) { [unowned self] (data, response, error) -> Void in
            self.didFinishRequest(request, response: response, data: data, error: error)
            }.resume()
    }
    
    private func didFinishRequest(request:NSURLRequest, response:NSURLResponse?, data:NSData?, error:NSError?) -> Void {
        //go to main thread if needed
        if !NSThread.isMainThread() {
            dispatch_async(dispatch_get_main_queue()) {
                [unowned self] in
                self.didFinishRequest(request, response: response, data: data, error: error)
            }
            return;
        }

        //handle network error
        if let error=error where self.isNetworkRelatedError(error) {
            //NSLog(@"+ Will Retry when network is up: %@", request.URL);
            
            //prepare to retry the request
            self.queuedRequests.insert(request, atIndex: 0)
            
            //wait for network coming back
            self.beginObservingReachabilityStatusForHost(request.URL!.host!);
            
            return;
        }
        
        //    NSLog(@"+ Did finish: %@", request.URL);
        
        self.queueRunning = false;
        
        //we are done with this one, saveList the array
        if(self.persistent) {
            self.saveList();
            //errors ignored
        }
        
        //tell our subclass and the delegate everything else
        self.dequeuedRequest(request, response: response, data: data, error: error)
        
        //go on
        self.doNextRequest()
    }
    
    private func isNetworkRelatedError(error:NSError) -> Bool {
        if (error.domain == NSURLErrorDomain){
            switch (error.code){
            case NSURLErrorInternationalRoamingOff, NSURLErrorCallIsActive, NSURLErrorDataNotAllowed, NSURLErrorNotConnectedToInternet:
                return true;
            default:
                return false;
            }
        }
        return false;
    }
    
    // MARK: hooks will inform the delegate but can be overriden for additional behavior

    func requestQueueWillStartRequest(request:NSURLRequest) -> Void {
        delegate?.requestQueueWillStartRequest?(self, request: request)
    }
    
    func requestQueueBeginWaitingForNetwork(host:String) -> Void {
        delegate?.requestQueueBeginWaitingForNetwork?(self, host: host)
    }
    func requestQueueEndWaitingForNetwork(host:String) -> Void {
        delegate?.requestQueueEndWaitingForNetwork?(self, host: host)
    }
    
    ///NOTE! called after done with a request for good or for bad -- this EXCLUDES network errors AND cancellations
    func dequeuedRequest(request:NSURLRequest, response:NSURLResponse?, data:NSData?, error:NSError?) {
        if let delegate = self.delegate {
            delegate.requestQueueDequeuedRequest(self, request: request, response: response, data: data, error: error)
        }
    }
    
    // MARK: serialization
    
    private func loadList() -> Bool {
        let docURL = self.URLForPersistentQueue;
        let br = true;
        
        //if it doesnt exist, quit
        //if it does, read it
        if NSFileManager.defaultManager().fileExistsAtPath(docURL.path!) {
            guard let codedData = NSData(contentsOfURL: docURL) else {
                return false
            }
            
            let unarchiver = NSKeyedUnarchiver(forReadingWithData: codedData)
            let array = unarchiver.decodeObjectForKey(DDRequestQueue.kDataKey)
            unarchiver.finishDecoding();
            
            guard let newRequests = array as? [NSURLRequest] else {
                return false
            }
            
            self.queuedRequests = newRequests
        }
        
        return br;
    }
    
    private func saveList() -> Bool {
        let array = self.queuedRequests;
        let docURL = self.URLForPersistentQueue;
        var br = true;
        
        //if the array is empty, delete the doc if it exists
        //if the array isnt empty, write it
        
        if(array.count == 0) {
            if NSFileManager.defaultManager().fileExistsAtPath(docURL.path!) {
                do {
                    try NSFileManager.defaultManager().removeItemAtURL(docURL)
                }
                catch {
                    br = false
                }
            }
        }
        else {
            let data = NSMutableData()
            let archiver = NSKeyedArchiver(forWritingWithMutableData: data)
            archiver.encodeObject(array, forKey: DDRequestQueue.kDataKey)
            archiver.finishEncoding()
            
            br = data.writeToURL(docURL, atomically: true)
        }
        return br;
    }
    
    lazy var URLForPersistentQueue:NSURL = {
        let library = try! NSFileManager.defaultManager().URLForDirectory(NSSearchPathDirectory.LibraryDirectory, inDomain: NSSearchPathDomainMask.UserDomainMask, appropriateForURL:nil, create: true)
        let appDir = library.URLByAppendingPathComponent(String(DDRequestQueue), isDirectory: true)
        let filename = "QueuedRequests-" + self.name;
        
        try! NSFileManager.defaultManager().createDirectoryAtURL(appDir, withIntermediateDirectories: true, attributes: nil)
        
        return appDir.URLByAppendingPathComponent(filename)
    }()
    
    // MARK: reachability
    
    private func beginObservingReachabilityStatusForHost(host:String) -> Void {
        let block: @convention(block)(SCNetworkReachabilityRef, SCNetworkReachabilityFlags, UnsafePointer<Void>) -> Void = { [unowned self]
            (reachability: SCNetworkReachabilityRef, flags: SCNetworkReachabilityFlags, data: UnsafePointer<Void>) in
            
            //the callback passes the wrong flags ;)
            var newflags = SCNetworkReachabilityFlags()
            let b = SCNetworkReachabilityGetFlags(reachability, &newflags)
            assert(b)
            
            self.hostDidBecomeReachable(host, reachable: newflags.contains(.Reachable))
        }
        
        let blockObject = imp_implementationWithBlock(unsafeBitCast(block, AnyObject.self))
        let fp = unsafeBitCast(blockObject, SCNetworkReachabilityCallBack.self)
        let utfString = (host as NSString).UTF8String
        
        if let reachabilityRef = SCNetworkReachabilityCreateWithName(kCFAllocatorDefault, utfString) where (host.characters.count > 0) {
            var context = SCNetworkReachabilityContext()
            
            if (SCNetworkReachabilitySetCallback(reachabilityRef, fp, &context)){
                if (!SCNetworkReachabilitySetDispatchQueue(reachabilityRef, dispatch_get_main_queue()) ){
                    // Remove our callback if we can't use the queue
                    SCNetworkReachabilitySetCallback(reachabilityRef, nil, nil);
                }
                self.currentReachability = reachabilityRef;
                
                self.requestQueueBeginWaitingForNetwork(host)
            }
        }
    }

    private func hostDidBecomeReachable(host:String, reachable:Bool) {
        if (reachable){
            self.requestQueueEndWaitingForNetwork(host)
            
            self.doNextRequest()
    
            SCNetworkReachabilitySetDispatchQueue(self.currentReachability, nil);
            //?!
//            CFRelease(self.currentReachability);
            self.currentReachability = nil;
        }
    }
}